apollo-federation/src/operation/mod.rs

```diff
@@ -2632,34 +2642,49 @@ impl SelectionSet {
+                    },
+                    SelectionValue::InlineFragment(fragment) => fragment
+                        .get_selection_set_mut()
+                        .add_at_path(path, selection_set),
+                    SelectionValue::FragmentSpread(_fragment) => {
+                        Err(FederationError::internal("add_at_path encountered a named fragment spread which should never happen".to_string()))
+                    }
+                }
+                }
                 let element = ele.rebase_on(&self.type_position, &self.schema)?;
                 let Some(sub_selection_type) = element.sub_selection_type_position()? else {
                     return Err(FederationError::internal("unexpected error: add_at_path encountered a field that is not of a composite type".to_string()));
                 };
-                let mut selection = Arc::make_mut(&mut self.selections)
-                    .entry(ele.key())
-                    .or_insert(|| {
+                let target = Arc::make_mut(&mut self.selections);
+                let selection = match target.get_mut(&ele.key()) {
+                    Some(selection) => selection,
+                    None => {
                         let unnecessary_directives = op_slice_condition_directives(path);
-                        Selection::from_element(
+                        let selection = Selection::from_element(
                             element,
                             // We immediately add a selection afterward to make this selection set
                             // valid.
                             Some(SelectionSet::empty(self.schema.clone(), sub_selection_type)),
                             Some(&unnecessary_directives),
-                        )
-                    })?;
-                match &mut selection {
-                    SelectionValue::Field(field) => match field.get_selection_set_mut() {
-                        Some(sub_selection) => sub_selection.add_at_path(path, selection_set)?,
-                        None => return Err(FederationError::internal("add_at_path encountered a field without a subselection which should never happen".to_string())),
-                    },
-                    SelectionValue::InlineFragment(fragment) => fragment
-                        .get_selection_set_mut()
-                        .add_at_path(path, selection_set)?,
-                    SelectionValue::FragmentSpread(_fragment) => {
-                        return Err(FederationError::internal("add_at_path encountered a named fragment spread which should never happen".to_string()));
+                        )?;
+                        target
+                            .entry(selection.key())
+                            .try_and_modify(|selection| {
+                                add_at_path_inner(selection, path, selection_set)
```

Comment:
wouldn't this effectively try to add at path twice? I think we just want to get a reference to the element here (or insert it) and then just do single call to `add_at_path_inner` at line #2687 (IMHO there is no need to have that extracted to a method)

Reply:
Ah, you're right. What should we do if we discover an existing entry? Do nothing and return the selection?

> IMHO there is no need to have that extracted to a method
Agreed, I only extracted it because I was using it in two locations.

Reply:
Yeah I think we just return the existing selection - then it should follow the same logic regardless whether we remove unneccessady directives or not (ie get an entry or try inserting new one)